### PR TITLE
New Prolog Cheat Sheet

### PR DESCRIPTION
**Pull request by arvchristos**
## Description of new Instant Answer, or changes

https://duck.co/ia/view/prolog_cheat_sheet

## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [ ] Instant Answer page is correctly filled out and contains:
    - The **Title** is appropriately named and formatted
    - The **IA topics** are present and appropriate
    - The **Description** is clear and coherent
    - **Source Name** exists if applicable
    - All **Example Queries** trigger on [Beta](https://beta.duckduckgo.com/)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/prolog_cheat_sheet

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by mintsoft** [comment]
Hmm we have a prolog cheatsheet inflight already, however nobody reviewed it https://github.com/duckduckgo/zeroclickinfo-goodies/pull/4065 this one looks more complete however

**Comment by arvchristos** [comment]
@mintsoft I didn't have a clue about this one. Hope this request doesn't appear anethical. 

_Sent from my Google Nexus 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by arvchristos** [comment]
@mintsoft What happens if both cheat sheets have the same alias??

**Comment by mintsoft** [comment]
Battle royale is the customary way of solving this problem...

In reality it depends, usually we'll pick the best one or whichever one is most active. This is larger than the other one so you're winning to begin with.

**Comment by mintsoft** [comment]
@moollaza this and https://github.com/duckduckgo/zeroclickinfo-goodies/pull/4065 have overlapping aliases. I'm thinking that this one is probably better anyway, is it possible to knock the SWI Prolog Cheatsheet off beta so we can test this one?

**Comment by mintsoft** [comment]
Testing locally this looks fine to me, I'm a little confused about the arithmetic terms syntax (`// /2` for example) but I'm assuming this is just me not being overly familiar with Prolog

**Comment by arvchristos** [comment]
@mintsoft This is the way to refer to prolog predicates. [foobar] /n is a 2 arity operator with name [foobar]. Thanks for the feedback

**Comment by mintsoft** [comment]
Ahh sounds reasonable then

**Comment by mintsoft** [comment]
@sahildua2305 @kirkins any problems you can see here?

**Comment by sahildua2305** [comment]
@mintsoft the only problem is with the overlapping aliases. Do we want to go ahead with this one and knock off the other one?

File: share/goodie/cheat_sheets/json/prolog.json [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+            {
+                "key": "told",
+                "val": "Close the current output stream"
+            }
+            
+        ],
+        "Pruning Operators": [
+            {
+                "key": "!\/0",
+                "val": "The cut is one of the Prolog operators related to the program control flow. It can be placed anywhere a goal can, and it is written as the predicate !/0. Technically, the cut commits the execution to all the choices made since the parent goal was unified with the head of the clause in which the cut appears. This means that all clauses below the one with the cut are discarded, as if they did not exist for this particular call (so they are not considered if the execution of the current clause fails, in which case the call to the predicate fails), and all the alternatives left by the execution of the goals at the left of the cut in that clause are also discarded. The goals at the right of the cut are executed normally (i.e., they can backtrack)."
```
Comment (sahildua2305) [inline share/goodie/cheat_sheets/json/prolog.json]: Too long, can we shorten this?

**Comment by mintsoft** [comment]
@sahildua2305 that's what I'm proposing, but I need @moollaza to remove the old one from demo. I can't do it I'm afraid

File: share/goodie/cheat_sheets/json/prolog.json (hunk 1 above)
Comment (arvchristos) [inline share/goodie/cheat_sheets/json/prolog.json]: I'm sure there is a shorter way to explain this specific one. Thanks for the feedback

Comment (arvchristos) [inline share/goodie/cheat_sheets/json/prolog.json]: @sahildua2305 Done

File: share/goodie/cheat_sheets/json/prolog.json [hunk 2]
```diff
@@ -165,7 +165,7 @@
...
             {
                 "key": "!\/0",
-                "val": "The cut is one of the Prolog operators related to the program control flow. It can be placed anywhere a goal can, and it is written as the predicate !/0. Technically, the cut commits the execution to all the choices made since the parent goal was unified with the head of the clause in which the cut appears. This means that all clauses below the one with the cut are discarded, as if they did not exist for this particular call (so they are not considered if the execution of the current clause fails, in which case the call to the predicate fails), and all the alternatives left by the execution of the goals at the left of the cut in that clause are also discarded. The goals at the right of the cut are executed normally (i.e., they can backtrack)."
+                "val": "A goal which always succeeds by preventing unwanted backtracking. All clauses below the one with the cut are discarded, as if they did not exist for this particular call and all the alternatives left by the execution of the goals at the left of the cut in that clause are also discarded. The goals at the right of the cut are executed normally"
```
Comment (mintsoft) [inline share/goodie/cheat_sheets/json/prolog.json]: This is probably still too long, can we call it the "cut" operator and leave it up to the user to understand the implications of using it?

**Comment by arvchristos** [comment]
@mintsoft I suppose yes. In fact the user who will use this cheatsheet should know the use of this predicate 

_Sent from my Google Nexus 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by cylgom** [comment]
ssdfsfsdfdfsf

**Comment by daxtheduck** [comment]
@arvchristos this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

**Comment by mintsoft** [comment]
Many thanks @arvchristos !

**Pull request merged**